### ISSUE TITLE
NG version status

### ISSUE THREAD

**Issue by raven2cz** [xinhaoyuan/layout-machi#13]
Hi Xinhao,
I already using ng branch few months and it works very nice. Do you have some plan to merge it to master?

I have second question to window switcher. I'm using your switcher, mod+/
Is there some possibility to implement standard replacement of windows in cycle without switcher editor. Like other layouts, it means cycle/rotate selected window for others selected regions.
The reason is that main features in NG branch nicely fill empty placements, and the layout is fulfilled  on whole screen, I need very often to swap windows between regions, not move or partial move, just shuffling the windows.

**Comment by basaran** [comment]
Hi, I haven't used the NG branch yet, but I just created a new module based on a few things I needed. It does move the windows using placement functions. If your layouts are not super complex, give it a try.

https://github.com/basaran/awesomewm-machina

**Comment by xinhaoyuan** [comment]
Hi raven2cz, Ng is merged into master now. For switching windows in all regions, you can try basaran's module, or use a general switcher, e.g. https://github.com/troglobit/awesome-switcher, or https://github.com/blueyed/awesome-cyclefocus.

**Comment by basaran** [comment]
Hi, thanks a lot for the new release. I just checked and my module is not working with it :) Due to the change of your internal regions function. I will try to migrate myself next week and update the machina module accordingly so it works with both versions. 

I also implemented a simple incremental and decremental region relocation without the use of internal the switcher which I will push that into the updated module as well. I think I can also make it directional if I add something to look around the active_region. But didn't have the time yet.

The new features of layout-machi look great, thank you again and have a nice week coming up.

**Comment by raven2cz** [comment]
Thanks! I will test it next week.

**Comment by raven2cz** [comment]
> Hi, thanks a lot for the new release. I just checked and my module is not working with it :) Due to the change of your internal regions function. I will try to migrate myself next week and update the machina module accordingly so it works with both versions.
> 
> I also implemented a simple incremental and decremental region relocation without the use of internal the switcher which I will push that into the updated module as well. I think I can also make it directional if I add something to look around the active_region. But didn't have the time yet.
> 
> The new features of layout-machi look great, thank you again and have a nice week coming up.

I will wait to your upgrade. It's good module. The main feature which I need is switch the position of each app between others app on the screen. The cyclefocus or switcher is not feature, which I need. I need swapping / shuffling active window to others place apps. This is standard behavior for others layout to have possibility change position.

I recorded video, what I mean.

https://user-images.githubusercontent.com/20504243/123328120-7bb24700-d53b-11eb-9081-b78725a2feb5.mp4

Just FYI: it seams that the basaran module is not ready for actual 4.4+ version, the `local keys = gears.table.join(...)`, doesn't work now. I have to modify it to `local keys = { .... }`, and finally use it by command:

`awful.keyboard.append_global_keybindings(machina.keys)`

**Comment by basaran** [comment]
what you are looking for is a functionality of the `auto` tiling window managers. layout-machi is a `manual` one, and the difference is, the way it splits the regions on a screen, and stacks the clients. Internal awesomewm `client:swap(c)` is not working with layout-machi. 

This should still be possible with some hacking though. I will look into it.

**Comment by basaran** [comment]
Try this:

```lua
   awful.key({modkey}, "F5", function ()
      local cltbl = awful.client.visible(client.focus.screen, true)
      local grect = require("gears.geometry").rectangle
      local geomtbl = {}

      for i,cl in ipairs(cltbl) do
         geomtbl[i] = cl:geometry()
      end

      local target = grect.get_in_direction("right", geomtbl, client.focus:geometry())

      tobe = cltbl[target]:geometry()
      is = client.focus:geometry()

      client.focus:geometry(tobe)
      cltbl[target]:geometry(is)
    end),
```

I will add at this to my module over the weekend with proper gap handling but for now this should work for you in your video layout without issues.

**Comment by basaran** [comment]
alright, I rolled out the new version of `awesomewm-machina`. It should work with both versions of layout-machi and it has quite a bit of added features which are probably buggy, but I will update it as I use it myself. It has the swapping feature implemented as well.

Many things changed, just remove the previous module and any configuration changes you made to your rc.lua and follow the instructions on the readme. You no longer should need to add the bindings manually.


https://user-images.githubusercontent.com/30809170/123538354-7d7a3700-d702-11eb-8db7-9114446389ee.mp4

**Comment by raven2cz** [comment]
I had time today for testing your new updates of awesomewm-machina, but it is stop working in awesome-git. I do not get any error response from AW. I tried to comment line:
capi.root.keys(awful.util.table.join(capi.root.keys(), table.unpack(bindings)))
but it doesn't help too. 

Do you have time to test with awesome-git (4.4+)? I will be week away, so I cannot test it in the next days...

**Comment by basaran** [comment]
when was the last time you cloned the repo, I had changed quite a few things and added new features, including real tabs. I'm thinking to add it as a release this week. But it should be all functional.

Please check the new readme, remove anything you added and follow the instructions on the readme. It should work no problem with awesome-git but I will test it too.

make sure you clone with:

```
git clone https://github.com/basaran/awesomewm-machina machina
```

and 

call with `local machina = require('machina')()`

**Comment by raven2cz** [comment]
I return back today. I test your last updates. I have clean installation of all libs and awesome-git. 

With awesome-git, there are big changes, and it seems that it doesn't work correctly.

First of all, I have to add the required deps to the all modules: methods. helpers, geoms, backham, etc.

There are missing required libs which caused nil exception:
```
local awful = require("awful")
local gears = require("gears")
local wibox = require("wibox")
local machi = require("layout-machi")
local beautiful = require("beautiful")
```
In my case, all are nil and it has to be required as local vars for usage with libs. So, I add required deps to methods, helpers, etc.

Second, the bindings is not automatically bind by calling init method. 

I have to manually add to the rc.lua this:

`awful.keyboard.append_global_keybindings(machina.bindings)`

After this, the machina is started. I tried bindings defined in the init.lua. 

I'm using mainly machi settings:  **dw88.**   it means dynamic grid 8x8

where the window can be expanded to the manual size like 2x2, 3x5 etc. 

After starting machina. The window sizes are lost information about remember window size, and have just 1x1 size only.
The memory of position and size is lost forever and it is not remember anymore.

Bindings:
```
awful.key({modkey}, "[", shuffle("backward")),
awful.key({modkey}, "]", shuffle("forward")),
```
This sometimes works for all windows, sometimes just for last two inserted, not for all in the screen. Unstable.

Next all bindings every time lost window size and set just 1x1 size, the window cannot be expand anymore. 

There is no bindings for swapping windows which I described. I tried to all bindings, but there is no binding for swap windows according to my video. And described example of code with F5 has nil exception, maybe still same problem because awesome-git, not 4.3 version.

It seems that the behavior of the previous awesome 4.3 and master branch is different and it needs different implementation/stabilization for actual master awesomewm version.

**Comment by basaran** [comment]
Welcome back)

Here are a few things:

Perhaps, in your rc.lua those libraries where required with local and they were not passed down to the machina module. We should have them required in the machina module to be on the safe side. (todo: I will add them, and make them local to machina module).

Most important bit is your layout dw88, this won't work the way you want it, because machina would consider each grid as it's own contrainer, and you would have very small containers individually. Perhaps I can dig more into machi internals to make it somehow work, but that's a big change. Try a layout like `332hv2`.

Shuffling logic also changed with the `tabs` I added to it. The first version was geometrically finding parent/child windows, which I suppose was able to find your expanded windows. This new version is going by regional indexes, so that's why it didn't work.

Those expanded regions are being reset, because machina is resetting them intentionally to keep the added tabs consistent.

Your keybinding update, I will add it to the docs (todo).

Swap bindings are in the init.lua

```lua
   awful.key({modkey, "Control"}, "j", shift_by_direction("left", "swap")),
   awful.key({modkey, "Control"}, "l", shift_by_direction("right", "swap")),
   awful.key({modkey, "Control"}, "k", shift_by_direction("down", "swap")),
   awful.key({modkey, "Control"}, "i", shift_by_direction("up","swap")),
   --▨ swap

   awful.key({modkey, "Control"}, "[", my_shifter("backward", "swap")),
   awful.key({modkey, "Control"}, "]", my_shifter("forward", "swap")),
   --▨ swap
```

So, in other words, all the new features I added has mostly steered away from your work flow. Try a layout like `332hv2` and see if things are working.

**Comment by basaran** [comment]
Here's the preview of directional swap:

https://user-images.githubusercontent.com/30809170/125209174-20839280-e265-11eb-8c40-182806b84aa6.mp4

**Comment by basaran** [comment]
Here's the preview of tabs:

https://user-images.githubusercontent.com/30809170/125209240-82dc9300-e265-11eb-9dcd-ad03219641be.mp4

**Comment by basaran** [comment]
Instead of the internal expansion, you can also use the expansion toggles I added. They will allow the regions (along with everything in that region to expand to the next region to the left or right or vertically. Hitting the same shortcut will toggle them back.


https://user-images.githubusercontent.com/30809170/125209522-4dd14000-e267-11eb-82a5-409576f06e89.mp4

**Comment by raven2cz** [comment]
I tested 332hv2 and it works according to your videos.

I found these issues:
- still missing required deps, it is better, you add some of them, but it is not enough. I have to add many of them to all lib files. And one need machi layout too.
```
local gears = require("gears")
local awful = require("awful")
local beautiful = require('beautiful')
local wibox = require("wibox")
local machi = require("layout-machi")
```
- tabs work correctly, but there is cross-feature error with not-active / active titlebars (git project "nice"). Screenshot attached.
![machina-tabs-cross-feature-titlebars](https://user-images.githubusercontent.com/20504243/125355450-5c058600-e365-11eb-83eb-d4e85aa0c9ba.png)

- bigger error: the attribute c.fullscreen stop working. c.maximized is working, but fullscreen window stop working, it is not possible to place window as fullscreen. The keybinding mod + f is not bind in machina/init.lua, but doesn't work. It seems that the fullscreen window is corrupted.

- FEATURE: I know that dw88 will be problem...It is my workflow which I'm using 60-70% for all my work now. Machina cannot be used for this grid layout. Mainly just swapping of windows is important. You will see in future...

- The remembered window size is lost. I don't know why, but machina kills important feature, that machi remembers the position and size of windows and if the application is closed and opened new instance again, the window is placed to specific size and placement. I'm using this all time, but machina kills this feature.

**Comment by basaran** [comment]
I haven't updated all the dependencies yet, and for your use case (including the nice package) perhaps it would be better and easier for you to just grab the snippet I sent before:

https://github.com/xinhaoyuan/layout-machi/issues/13#issuecomment-868929662

Just modify it for your need add the awful dependency, and it should allow you to swap clients.

**Comment by raven2cz** [comment]
> I haven't updated all the dependencies yet, and for your use case (including the nice package) perhaps it would be better and easier for you to just grab the snippet I sent before:
> 
> [#13 (comment)](https://github.com/xinhaoyuan/layout-machi/issues/13#issuecomment-868929662)
> 
> Just modify it for your need add the awful dependency, and it should allow you to swap clients.

yes, I already try this function. But cltbl and target variables are nil. Maybe `client.focus` or `awful.client.visible()` works different in 4.4 now?

**Comment by xinhaoyuan** [comment]
closing this as ng is already merged.
